File: Practical_Task_3/Correct_Ones/APIAccess.php
<?php

class APIAccess {
    public function main() {
        $apiKey = getenv('API_KEY');  // Secure environment variable
        if (!$apiKey) {
            error_log("API Key is not set.");
            die("Internal server error. Please try again later.");  // Avoid exposing internal details to the user
        }

        $userApiKey = $this->getUserApiKey();

        if ($this->authenticate($apiKey, $userApiKey)) {
            echo "Access granted.\n";
        } else {
            echo "Access denied.\n";
        }
    }

    private function authenticate($apiKey, $userApiKey) {
        return hash_equals($apiKey, $userApiKey);  // Constant-time comparison to prevent timing attacks
    }

    private function getUserApiKey() {
        $userApiKey = $_POST['apiKey'] ?? null;  // Use POST method for secure API key transmission
        if (empty($userApiKey)) {
            error_log("User API Key is missing.");
            die("Internal server error. Please try again later.");
        }

        return htmlspecialchars(strip_tags($userApiKey));  // Proper sanitation to prevent XSS and injection attacks
    }
}

$apiAccess = new APIAccess();
$apiAccess->main();

File: Practical_Task_3/Correct_Ones/Java/APIAccess.java
package Practical_Task_3.Correct_Ones.Java;
// In terminal:

// export API_KEY="1234567890"


import java.util.Scanner;

public class APIAccess {
    public static void main(String[] args) {
        try {
            String apiKey = System.getenv("API_KEY"); // Retrieve API key from environment variable
            if (apiKey == null) {
                throw new RuntimeException("API_KEY environment variable not set.");
            }
            String userApiKey = getUserApiKey();

            if (authenticate(apiKey, userApiKey)) {
                System.out.println("Access granted.");
            } else {
                System.out.println("Access denied.");
            }
        } catch (Exception e) {
            System.err.println("Error occurred: " + e.getMessage());
        }
    }

    private static boolean authenticate(String apiKey, String userApiKey) {
        return apiKey != null && apiKey.equals(userApiKey);
    }

    private static String getUserApiKey() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter your API key: ");
        return scanner.nextLine();
    }
}

